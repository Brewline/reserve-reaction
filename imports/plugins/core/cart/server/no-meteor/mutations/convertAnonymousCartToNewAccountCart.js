import Random from "@reactioncommerce/random";
import ReactionError from "@reactioncommerce/reaction-error";
import { Cart as CartSchema } from "/imports/collections/schemas";
import appEvents from "/imports/node-app/core/util/appEvents";

/**
 * @summary Copy items from an anonymous cart into a new account cart, and then delete the
 *   anonymous cart.
 * @param {String} accountId The account ID to associate with the new account cart
 * @param {Object} anonymousCart The anonymous cart document
 * @param {Object} anonymousCartSelector The MongoDB selector for the anonymous cart
 * @param {MongoDB.Collection} Cart The Cart collection
 * @param {String} shopId The shop ID to associate with the new account cart
 * @return {Object} The new account cart
 */
export default async function convertAnonymousCartToNewAccountCart({
  accountId,
  anonymousCart,
  anonymousCartSelector,
  Cart,
  shopId,
  userId
}) {
  const createdAt = new Date();
  const currencyCode = anonymousCart.currencyCode || "USD";
  const newCart = {
    _id: Random.id(),
    accountId,
    anonymousAccessToken: null,
    currencyCode,
    createdAt,
    items: anonymousCart.items,
    shopId,
    updatedAt: createdAt,
    workflow: {
      status: "new"
    }
  };

  CartSchema.validate(newCart);

  const { result } = await Cart.insertOne(newCart);
  if (result.ok !== 1) throw new ReactionError("server-error", "Unable to create account cart");

  await appEvents.emit("afterCartCreate", {
    cart: newCart,
    createdBy: userId
  });

  const { deletedCount } = await Cart.deleteOne(anonymousCartSelector);
  if (deletedCount === 0) throw new ReactionError("server-error", "Unable to delete anonymous cart");

  return newCart;
}
