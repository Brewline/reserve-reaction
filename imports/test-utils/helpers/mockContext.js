const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
