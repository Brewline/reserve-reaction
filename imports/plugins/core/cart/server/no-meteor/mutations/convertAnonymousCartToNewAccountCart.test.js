import mockContext from "/imports/test-utils/helpers/mockContext";
import convertAnonymousCartToNewAccountCart from "./convertAnonymousCartToNewAccountCart";

const { Cart } = mockContext.collections;
const currencyCode = "GBP";
const accountId = "accountId";
const anonymousCartSelector = { _id: "123" };
const shopId = "shopId";
const items = [
  {
    _id: "CartItemID",
    addedAt: new Date("2018-01-01T00:00:00.000"),
    createdAt: new Date("2018-01-01T00:00:00.000"),
    productId: "productId",
    quantity: 1,
    shopId: "shopId",
    title: "TITLE",
    updatedAt: new Date("2018-01-01T00:00:00.000"),
    variantId: "variantId",
    isTaxable: false,
    priceWhenAdded: {
      amount: 9.99,
      currencyCode: "USD"
    }
  }
];

test("inserts a cart with the existing cart's items and returns it", async () => {
  Cart.insertOne.mockReturnValueOnce(Promise.resolve({ result: { ok: 1 } }));

  const result = await convertAnonymousCartToNewAccountCart({
    accountId,
    anonymousCart: {
      currencyCode,
      items
    },
    anonymousCartSelector,
    Cart,
    shopId
  });

  const newCart = {
    _id: jasmine.any(String),
    accountId,
    anonymousAccessToken: null,
    currencyCode,
    createdAt: jasmine.any(Date),
    items,
    shopId,
    updatedAt: jasmine.any(Date),
    workflow: {
      status: "new"
    }
  };

  expect(Cart.insertOne).toHaveBeenCalledWith(newCart);

  expect(Cart.deleteOne).toHaveBeenCalledWith(anonymousCartSelector);

  expect(result).toEqual(newCart);
});

test("throws if insertOne fails", async () => {
  Cart.insertOne.mockReturnValueOnce(Promise.resolve({ result: { ok: 0 } }));

  const promise = convertAnonymousCartToNewAccountCart({
    accountId,
    anonymousCart: {
      currencyCode,
      items
    },
    anonymousCartSelector,
    Cart,
    shopId
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if deleteOne fails", async () => {
  Cart.insertOne.mockReturnValueOnce(Promise.resolve({ result: { ok: 1 } }));
  Cart.deleteOne.mockReturnValueOnce(Promise.resolve({ deletedCount: 0 }));

  const promise = convertAnonymousCartToNewAccountCart({
    accountId,
    anonymousCart: {
      currencyCode,
      items
    },
    anonymousCartSelector,
    Cart,
    shopId
  });

  return expect(promise).rejects.toThrowErrorMatchingSnapshot();
});
