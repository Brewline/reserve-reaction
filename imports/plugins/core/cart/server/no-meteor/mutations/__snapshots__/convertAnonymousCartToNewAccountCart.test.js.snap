// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if deleteOne fails 1`] = `"Unable to delete anonymous cart"`;

exports[`throws if insertOne fails 1`] = `"Unable to create account cart"`;
